docs/source/diagnostics.rst

```diff
@@ -0,0 +1,86 @@
+Diagnostics
+===========
+
+Profiling parallel code can be tricky. ``dask.diagnostics`` provides
+functionality to aid in profiling and inspecting dask graph execution.
+
+
+Scheduler Callbacks
+-------------------
+
+Schedulers based on ``dask.async.get_async`` (currently
+``dask.async.get_sync``, ``dask.threaded.get``, and
+``dask.multiprocessing.get``) accept two callbacks, allowing for inspection of
+dask execution. The callbacks are:
+
+1. ``start_callback(self, key, dask, state)``
+    Callback ran every time a new task is started. Receives the key of the
+    task to be run, the dask, and the scheduler state. The final callback
+    receives ``None`` instead of a key, as no new tasks were added at that
+    tick.
+2. ``end_callback(self, key, dask, state, id)``
+    Callback ran every time a task is finished. Receives the key of the
+    task to be run, the dask, the scheduler state, and the id of the worker
+    that ran the task. The final callback receives ``None`` none for both key
+    worker id, as no new tasks were finished at that tick.
+
+
+ Callbacks for common use cases are provided in ``dask.diagnostics``.
+
+
+Profiler
+--------
+
+The ``Profiler`` class profiles dask execution at the task level. To use,
+create a profiler from a scheduler ``get`` function:
+
+.. code-block:: python
+
+    >>> from dask.array.core import get
+    >>> from dask.diagnostics import Profiler
+    >>> array_profiler = Profiler(get)
+
+The ``get`` method of the profiler then works like a normal scheduler, but
+records the following information for each task during execution:
+
+1. Key
+2. Task
+3. Start time in seconds since the epoch
+4. Finish time in seconds since the epoch
+5. Worker id
+
+.. code-block:: python
+
+    >>> import dask.array as da
+    >>> a = da.random.random(size=(1000,10000), chunks=(1000,1000))
+    >>> q, r = da.linalg.qr(a)
```

Comment:
`da.linalg.qr` only works on tall-and-skinny matrices and so this will generate an error.  It would be ok if you were to transpose the size though.